ompi/mpi/c/init_thread.c.in

```diff
@@ -39,22 +40,32 @@
 PROTOTYPE ERROR_CLASS init_thread(INT_OUT argc, ARGV argv, INT required,
                                   INT_OUT provided)
 {
-    int err, safe_required = MPI_THREAD_SERIALIZED;
+    int err, err_arg_required = false, safe_required = MPI_THREAD_SERIALIZED;
     char *env;
 
     ompi_hook_base_mpi_init_thread_top(argc, argv, required, provided);
 
     /* Detect an incorrect thread support level, but dont report until we have the minimum
      * infrastructure setup.
      */
-    if( (MPI_THREAD_SINGLE == required) || (MPI_THREAD_SERIALIZED == required) ||
-        (MPI_THREAD_FUNNELED == required) || (MPI_THREAD_MULTIPLE == required) ) {
+    err_arg_required = (required != MPI_THREAD_SINGLE && required != MPI_THREAD_FUNNELED &&
+                        required != MPI_THREAD_SERIALIZED && required != MPI_THREAD_MULTIPLE);
+    if (!err_arg_required) {
+        safe_required = required;
+    }
 
-        if (NULL != (env = getenv("OMPI_MPI_THREAD_LEVEL")))  {
-            safe_required = atoi(env);
-        }
-        else {
-            safe_required = required;
+    /* check for environment overrides for required thread level.  If
+     * there is, check to see that it is a valid/supported thread level.
+     *  if it is lower than argument required, set it anyway, program can
+     * check provided == required to verify that required has been ignored
+     */
+    if (NULL != (env = getenv("OMPI_MPI_THREAD_LEVEL"))) {
+        int env_required = atoi(env);
+        err_arg_required = err_arg_required ||
```

Comment:
```suggested
env_arg_required |= ...etc
```

Reply:
For now, the valid values are still [0-3] (because our internal ABI is not the same as the standardized ABI (yet?), I'll add this consideration to the comment requested by Jeff so we don't forget when the time comes.

Reply:
FWIW: we plan to support both the existing OMPI ABI and the official MPI Forum MPI ABI.  Hence, we'll probably want code that can handle both situations.

Reply:
the important part is that we don't want the users to have to enter the ABI values for the environment variable for the thread level. We want this to remain indexed as it is today.

Reply:
can we defer this to when we implement ABI compat?

Reply:
It might behoove us to do this now, so that it can be in 6.0.0 -- i.e., at a major version change.  ABI may or may not make it into 6.0.0 (e.g., it won't be ready until -- at the earliest -- the end of this summer, and we may have already released 6.0.0 by then, in which case, ABI would be in 6.1.0).

It can't be that hard to implement to also check for string values.  I'd suggest the following case-insensitive values:

* `mpi_thread_single` and `single`
* `mpi_thread_serialized` and `serialized`
* `mpi_thread_funneled` and `funneled`
* `mpi_thread_multiple` and `multiple`

Reply:
Its done